### PR TITLE
Add FDT memory and reserved memory node

### PR DESCRIPTION
**Pull request by gdong1**
PlatformInit need report memory related information using these nodes before handling control to universal payload.

### CODE REVIEW

File: docs/chapter4-payload-handoff-format.rst [hunk 1]
```diff
@@ -35,3 +35,139 @@ architecture, either 32 or 64 bits.
 Many of these sections reference the Devicetree Specification [DTspec]_.
 
 TBD (Nodes)
+
+4.2.8 Node: /memory (R)
+~~~~~~~~~~~~~~~~~~~~~~~
+
+The memory node is required to describe the physical-memory layout for the system. If a system has multiple ranges of memory, multiple memory nodes can be created, or the ranges can be specified in the reg property of a single memory node. See the Devicetree Specification ("/memory node" section) for details.
```
Comment (sjg20) [inline docs/chapter4-payload-handoff-format.rst:42]: Can you please format to 80cols?

File: docs/chapter4-payload-handoff-format.rst [hunk 2]
```diff
@@ -35,3 +35,139 @@ architecture, either 32 or 64 bits.
...
+
+.. tabularcolumns:: | p{3cm} p{0.75cm} p{2cm} p{9.5cm} |
+.. table:: Node: memory@base_address
+
+  ==================== ========= =============== ======================================================
+   Property Name        Usage     Value Type      Definition
+  ==================== ========= =============== ======================================================
+   device_type           R       string          Value shall be �memory�
```
Comment (sjg20) [inline docs/chapter4-payload-handoff-format.rst:50]: What is the strange character around 'memory'?

Perhaps use `'memory'` ?

File: docs/chapter4-payload-handoff-format.rst [hunk 3]
```diff
@@ -35,3 +35,139 @@ architecture, either 32 or 64 bits.
...
+   reg                   R       prop-encoded-   Specify system memory region range. Consists of an
+                                 array           arbitrary number of address and size pairs that specify
+                                                 the physical address and size of the memory ranges.
+   initial-mapped-       O       prop-encoded-   <u64 - effective address, u64 - physical address,
```
Comment (sjg20) [inline docs/chapter4-payload-handoff-format.rst:54]: missing text after initial-mapped-

File: docs/chapter4-payload-handoff-format.rst [hunk 4]
```diff
@@ -35,3 +35,139 @@ architecture, either 32 or 64 bits.
...
+        device_type = "memory";
+        reg = < 0x000000001 0x00000000 0x00000001 0x00000000 >;
+        attribute = < 0x0000000000000000 >;
+    };
+
+
+4.2.9 Node: reserved-memory (R)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The /reserved-memory node provides information about memory which is not available for use. Payload shall exclude reserved memory from normal usage. One can create child nodes describing particular reserved (excluded from normal use) memory regions. Such memory regions are usually designed for the special usage by various device drivers.
```
Comment (sjg20) [inline docs/chapter4-payload-handoff-format.rst:106]: samd comment re wrapping

File: docs/chapter4-payload-handoff-format.rst [hunk 5]
```diff
@@ -35,3 +35,139 @@ architecture, either 32 or 64 bits.
...
+   Property Name      Usage     Value Type       Definition
+  ================ =========== ================ ======================================================
+    reg                R        u32 / 64 array   Specify memory region of reserved memory
+    no-map             O        boolean          If present, indicates the operating system must not 
+                                                 create a virtual mapping of the region
+    compatible         O        string list
+  ================ =========== ================ ======================================================
+
+Child Nodes for Payload Memory Types:
+One of the example usage is that UEFI Payload will be aware of where pre-installed acpi tables or NVS regions are and will adopt them for supporting additional ACPI table installation from payload phase. Those boot-code/boot-data regions from PlatformInit will be reported to UEFI OS as usable memory without waste. Similarly runtime-code/runtime-data may be provided by PlatformInit for supporting UEFI runtime services that will be utilized by UEFI OS. All of the below are optional and can be skipped if unsupported by the platform.
```
Comment (sjg20) [inline docs/chapter4-payload-handoff-format.rst:134]: wrapping again

File: docs/chapter4-payload-handoff-format.rst [hunk 6]
```diff
@@ -35,3 +35,139 @@ architecture, either 32 or 64 bits.
...
+Child Nodes for Payload Memory Types:
+One of the example usage is that UEFI Payload will be aware of where pre-installed acpi tables or NVS regions are and will adopt them for supporting additional ACPI table installation from payload phase. Those boot-code/boot-data regions from PlatformInit will be reported to UEFI OS as usable memory without waste. Similarly runtime-code/runtime-data may be provided by PlatformInit for supporting UEFI runtime services that will be utilized by UEFI OS. All of the below are optional and can be skipped if unsupported by the platform. 
+
+
+  ================== ======================================================
+  Compatible string       Description
+  ================== ======================================================
+    acpi              Memory that holds ACPI tables
+    acpi-nvs          ACPI NVS buffer information
+    boot-code         Memory that holds firmware phase drivers and will be released to the OS when firmware boot phase finishes (for example with a UEFI payload when ExitBootService signaled).
```
Comment (sjg20) [inline docs/chapter4-payload-handoff-format.rst:142]: and again

File: docs/chapter4-payload-handoff-format.rst [hunk 7]
```diff
@@ -35,3 +35,139 @@ architecture, either 32 or 64 bits.
...
+    boot-data         Memory that holds firmware phase data consumed by drivers and will be released to the OS when firmware boot phase finishes (for example with a UEFI payload when ExitBootService signaled).
+    runtime-code      Runtime service code memory region which will be utilized by OS runtime service
```
Comment (sjg20) [inline docs/chapter4-payload-handoff-format.rst:144]: How about 'used' instead of 'utilized' ?

**Review by sjg20 — APPROVED** [review]
This looks good to me, except for some nits

File: docs/chapter4-payload-handoff-format.rst (hunk 1 above)
Comment (gdong1) [reply docs/chapter4-payload-handoff-format.rst:42]: sure. it is fixed in the updated PR.

File: docs/chapter4-payload-handoff-format.rst (hunk 7 above)
Comment (gdong1) [reply docs/chapter4-payload-handoff-format.rst]: sure. it was fixed.

File: docs/chapter4-payload-handoff-format.rst (hunk 6 above)
Comment (gdong1) [reply docs/chapter4-payload-handoff-format.rst]: sure. it was fixed.

File: docs/chapter4-payload-handoff-format.rst (hunk 5 above)
Comment (gdong1) [reply docs/chapter4-payload-handoff-format.rst]: sure. it was fixed.

File: docs/chapter4-payload-handoff-format.rst (hunk 4 above)
Comment (gdong1) [reply docs/chapter4-payload-handoff-format.rst]: sure. it was fixed.

File: docs/chapter4-payload-handoff-format.rst (hunk 3 above)
Comment (gdong1) [reply docs/chapter4-payload-handoff-format.rst]: sure. it was fixed.

File: docs/chapter4-payload-handoff-format.rst (hunk 2 above)
Comment (gdong1) [reply docs/chapter4-payload-handoff-format.rst]: sure. it was fixed.

**Review by ChaselChiu — APPROVED** [review]

File: docs/chapter4-payload-handoff-format.rst [hunk 8]
```diff
@@ -35,3 +35,173 @@ architecture, either 32 or 64 bits.
 Many of these sections reference the Devicetree Specification [DTspec]_.
 
 TBD (Nodes)
+
+4.2.8 Node: /memory (R)
```
Comment (LeanSheng) [inline docs/chapter4-payload-handoff-format.rst:39]: 4.2.10

File: docs/chapter4-payload-handoff-format.rst [hunk 9]
```diff
@@ -35,3 +35,173 @@ architecture, either 32 or 64 bits.
...
+  ================ =========== ============ =====================================
+   #address-cells       R
+   #size-cells          R
+   ranges
+  ================ =========== ============ =====================================
+
+Child Node:
+
+.. tabularcolumns:: | p{3cm} p{0.75cm} p{2cm} p{9.5cm} |
+.. table:: Node: name@xxx
```
Comment (LeanSheng) [inline docs/chapter4-payload-handoff-format.rst:138]: .. table:: Node: name@<address>

File: docs/chapter4-payload-handoff-format.rst [hunk 10]
```diff
@@ -35,3 +35,173 @@ architecture, either 32 or 64 bits.
...
+~~~~~~~~~~~~~~~~~~~~~~~
+
+The memory node is required to describe the physical-memory layout for the
+system. If a system has multiple ranges of memory, multiple memory nodes can
+be created, or the ranges can be specified in the reg property of a single
+memory node. See the Devicetree Specification ("/memory node" section) for
+details.
+
+.. tabularcolumns:: | p{3cm} p{0.75cm} p{2cm} p{9.5cm} |
+.. table:: Node: memory@base_address
```
Comment (LeanSheng) [inline docs/chapter4-payload-handoff-format.rst:49]: `.. table:: Node: memory@<base_address>`

File: docs/chapter4-payload-handoff-format.rst [hunk 11]
```diff
@@ -35,3 +35,173 @@ architecture, either 32 or 64 bits.
...
+        device_type = "memory";
+        reg = < 0x000000001 0x00000000 0x00000001 0x00000000 >;
+        attribute = < 0x0000000000000000 >;
+    };
+
+
+4.2.9 Node: reserved-memory (R)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The /reserved-memory node provides information about memory which is not available
```
[the remark this answers is not in the document]
Comment (ChaselChiu) [reply docs/chapter4-payload-handoff-format.rst:118]: looks good.

[the remark this answers is not in the document]
Comment (dhaval-rivos) [reply docs/chapter4-payload-handoff-format.rst:118]: lgtm

[the remark this answers is not in the document]
Comment (LeanSheng) [reply docs/chapter4-payload-handoff-format.rst:118]: will mark as resolved after @gdong1 pushed the update.

**Review by dhaval-rivos — APPROVED** [review]

**Comment by LeanSheng** [comment]
would you mind to also edit here to add the 2 nodes?
https://github.com/UniversalPayload/spec/blob/main/docs/revhistory.rst
if you are bumping to v0.9.1 would also need changes here:
https://github.com/UniversalPayload/spec/blob/0b030769621c9667949fc2b294abd1e3939bafe1/docs/conf.py#L38

File: docs/chapter4-payload-handoff-format.rst (hunk 8 above)
Comment (gdong1) [reply docs/chapter4-payload-handoff-format.rst]: done

File: docs/chapter4-payload-handoff-format.rst [hunk 12]
```diff
@@ -35,3 +35,173 @@ architecture, either 32 or 64 bits.
...
+        attribute = < 0x0000000000000001 >;
+    };
+    memory@1000000000 {
+        device_type = "memory";
+        reg = < 0x000000001 0x00000000 0x00000001 0x00000000 >;
+        attribute = < 0x0000000000000000 >;
+    };
+
+
+4.2.9 Node: reserved-memory (R)
```
[the remark this answers is not in the document]
Comment (gdong1) [reply docs/chapter4-payload-handoff-format.rst]: done

File: docs/chapter4-payload-handoff-format.rst (hunk 11 above)
[the remark this answers is not in the document]
Comment (gdong1) [reply docs/chapter4-payload-handoff-format.rst]: done

File: docs/chapter4-payload-handoff-format.rst (hunk 9 above)
Comment (gdong1) [reply docs/chapter4-payload-handoff-format.rst]: done

File: docs/chapter4-payload-handoff-format.rst (hunk 10 above)
Comment (gdong1) [reply docs/chapter4-payload-handoff-format.rst]: done

**Comment by gdong1** [comment]
> ld also need changes h

done

**Comment by LeanSheng** [comment]
Would you mind to bump the version here as well?

https://github.com/UniversalPayload/spec/blob/0b030769621c9667949fc2b294abd1e3939bafe1/docs/conf.py#L38

**Comment by gdong1** [comment]
> Would you mind to bump the version here as well?
> 
> https://github.com/UniversalPayload/spec/blob/0b030769621c9667949fc2b294abd1e3939bafe1/docs/conf.py#L38

done

**Review by LeanSheng — APPROVED** [review]

**Pull request merged**